lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+      class HalfEdge
+        include Comparable
+
+        # Creates 2 half edges from an edge.
+        # They will be assigned as each other's twins.
+        # The Half Edges will be returned in the order of points
+        # of the edge (start, end).
+        #
+        # @param edge [RGeo::Cartesian::Segment]
+        #
+        # @return [Array]
+        def self.from_edge(edge)
+          # TODO: should we raise an error if the edge is degenerate?
+          e1 = new(edge.s)
+          e2 = new(edge.e)
+
+          e1.twin = e2
+          e2.twin = e1
+          [e1, e2]
+        end
+
+        def initialize(origin)
+          @origin = origin
+          @twin = nil
+          @next = nil
+          @prev = nil
+        end
+        attr_reader :origin
+        attr_accessor :twin, :next, :prev
+
+        # HalfEdges will be sorted around their shared vertex
+        # in a CW fashion. This means that face interiors will be
+        # a CCW.
+        def <=>(other)
+          angle <=> other.angle
+        end
+
+        # Will keep following next in a half-edge until it returns to itself
+        # or hits a half-edge without a next edge.
+        #
+        # @return [Array]
+        def each
+          hedges = []
+          yield(self) if block_given?
+          hedges << self
+
+          n = self.next
+          until n.eql?(self) || n.nil?
+            yield(n) if block_given?
+            hedges << n
+            n = n.next
+          end
+          hedges
+        end
+
+        # Return the destination of the half edge
+        #
+        # @return [RGeo::Feature::Point]
+        def destination
+          twin.origin
+        end
+
+        # Compute the angle from the positive x-axis.
+        # Used for sorting at each node.
+        #
+        # @return [Float]
+        def angle
+          @angle ||= Math.atan2(destination.y - origin.y, destination.x - origin.x)
+        end
+
+        def inspect
+          "#<#{self.class}:0x#{object_id.to_s(16)} #{self}>"
+        end
+
+        def to_s
+          dst = twin.nil? ? nil : destination
+          pr = prev.nil? ? nil : prev.origin
+          n = @next.nil? ? nil : @next.origin
+          "HalfEdge(#{origin}, #{dst}), Prev: #{pr},  Next: #{n}"
+        end
+      end
+
+      # A Doubly Connected Edge List (DCEL) implementation of a Planar Graph.
+      # It represents geometries as vertices and half-edges.
+      #
+      # It includes an incident_edges hash that maps vertices to an array
+      # of half-edges whose origins are that vertex.
+      #
+      # Upon instantiation, the graph will compute the intersections using
+      # the SweeplineIntersector, populate the incident_edges map, and
+      # link all cyclic edges.
+      class PlanarGraph
+        # Create a new PlanarGraph
+        #
+        # @param edges [Array<RGeo::Cartesian::Segment>] of Segments
+        def initialize(edges = nil)
+          edges = [] if edges.nil?
```

Comment:
why not
```suggestion
        def initialize(edges = [])
```

Reply:
I know JavaScript you're not really supposed to use objects as a default value since if they're mutated, the function may still reference that object. Not sure if it's the case with Ruby as well.